graphbolt/src/cnumpy.h

```diff
@@ -0,0 +1,117 @@
+/**
+ *  Copyright (c) 2023 by Contributors
+ * @file cnumpy.h
+ * @brief Numpy File Fetecher class.
+ */
+
+#include <fcntl.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <torch/script.h>
+
+#ifdef __linux__
+#include <liburing.h>
+#include <unistd.h>
+#endif
+
+#include <cassert>
+#include <cstdio>
+#include <cstring>
+#include <fstream>
+#include <iostream>
+#include <string>
+
+namespace graphbolt {
+namespace storage {
+
+/**
+ * @brief Disk Numpy Fetecher class.
+ */
+class OnDiskNpyArray : public torch::CustomClassHolder {
+ public:
+  /** @brief Default constructor. */
+  OnDiskNpyArray() = default;
+
+  /**
+   * @brief Constructor with given file path and data type.
+   * @param path Path to the on disk numpy file.
+   * @param dtype Data type of numpy array.
+   *
+   * @return OnDiskNpyArray
+   */
+  OnDiskNpyArray(std::string filename, torch::ScalarType dtype);
+
+  /** @brief Create a disk feature fetcher from numpy file. */
+  static c10::intrusive_ptr<OnDiskNpyArray> Create(
+      std::string path, torch::ScalarType dtype);
+
+  /** @brief Deconstructor. */
+  ~OnDiskNpyArray();
+
+  /**
+   * @brief Parses the header of a numpy file to extract feature information.
+   *
+   * Numpy files have a specific binary storage format. This method parses the
+   * header of the numpy file. It first extracts the header information of the
+   * numpy file by string manipulation, ending with a character "\n".
+   * The header contains three descriptors: 'descr' for data format description
+   * 'fortran_order' for row or column order preference, and 'shape' for data
+   * shape. The data format description includes the endianness, data type, and
+   * data length, e.g. '<i8' denotes little-endian storage, integer type, and
+   * 8-byte size (64 bit integer).
+   *
+   * For example, for a tensor `A` having 2 features with following entries:
+   * [[1, 2, 3], [4, 5, 6]]
+   * When `A` is saved as an numpy file, the data is stored as:
+   *   �NUMPY'descr': '<i8', 'fortran_order': False, 'shape': (2, 3), }\n
```

Comment:
what is the first?

Reply:
It seems to be a little binary data that cannot be represented.